histoqc/BaseImage.py

```diff
@@ -110,6 +110,15 @@ def getBestLevelForDownsample(self, downsample_factor: float) -> Tuple[int, bool
         else:
             return (osh.get_best_level_for_downsample(downsample_factor), False)
 
+
+    @staticmethod
```

Comment:
maybe this goes in the WSI handle file?

as well, I'm wondering if these should be an if statement or a try/catch statement?

or maybe a try catch should be in place around the if? 

i guess in this case its obvious which file type to use? might need to check with petros

Reply:
I still think about this part. We probably should use factory pattern + dynamic import to implement how to create subclass for ABC class.

Reply:
@nanli-emory @choosehappy 
I agree.
If I understand correctly about what you want here:
(1) avoid importing each subclass handles because not all might be installed on the client's machine.
(2) dynamically create the corresponding subclass handle based on what's installed or any input arguments.

So herein you may have the below options:
(1) create a factory class (or make it extending Protocol for Python 3.10 and above) that implements the relevant APIs and the builder:
`
class HandleBuilder:
    @staticmethod
    def build(params) -> WSIImageHandle:
        ... # return different handle types based on 
`

(2) Optionally, rather than just making a HandleBuilder, make the supposed factory builder a higher level of abstraction in some way close to the skeletal pattern. Herein your WSIImageHandle would be the abstract class to manage how to open the image resources and compute corresponding attributes by the data read accordingly.
You can create another interface (e.g., `class ImageCore` )that 
    --- (1)  associate a WSIImageHandle as a private attribute.
    --- (2) provide all APIs of WSIImageHandle by returning the results of the associated WSIImageHandle 
    --- (3) implement the factory builder function above as a static method or class method

By doing this the advantage is, you can move certain behavior like "get_best_level_for_downsample" from BaseImage to `ImageCore` and therefore your BaseImage can focus on carrying data/mask/results between each module. 

(3) Simply use a function as the factory builder.
![image](https://github.com/choosehappy/HistoQC/assets/10933828/bdaf0ba5-4489-4986-b9f3-2b9e6260f410)

Reply:
"Note that essentially you want to hide the details of all WSIImageHandle's subclasses from BaseImage as HistoQC's pipeline itself shouldn't care about it anyway, therefore the import of all subclasses should be avoided in the BaseImage class."

Yes, exactly this. this is what it looks like, which also addresses the first comment in this review and last comment:

```python
#---main.py
from baseclass import AbstracthbaseClass

user_list=["openslide","dicom"]

for types in user_list:
    ci = AbstracthbaseClass.returnClass(types)
    print(ci)
    print(ci.has_bounding_box)
from abc import ABC, abstractmethod

#---baseclass.py
class AbstracthbaseClass(ABC):


    @staticmethod
    def returnClass(type):
        if type=="openslide":
            print("c1")
            from c1class import c1
            return c1(type)
        else:
            print("c2")
            from c2class import c2
            return c2(type)

    @property
    @abstractmethod
    def has_bounding_box(self):
        ...

#--- c1class.py
from baseclass import AbstracthbaseClass

class c1(AbstracthbaseClass):
    def __init__(self, fname):
        self.fname = fname
        print("c1 init")

    @property
    def has_bounding_box(self):
        print("c1 bounding box")
        return self.fnamefrom baseclass import AbstracthbaseClass

#----c2class.py
class c2(AbstracthbaseClass):
    def __init__(self, fname):
        self.fname = fname
        print("c2 init")

    @property
    def has_bounding_box(self):
        print("c2 bounding box")

```

Reply:
Hi @choosehappy, I Changed how to load and create subclasses for WSIImageHandle. But do we ready need to have "handles" parameters for ImageBase? 
Currently, **create_wsi_handle** function will auto-determine the file type and import the subclass if needed.